### ISSUE TITLE
Generator function typed with non optional parameter, throws error TS2345

### ISSUE THREAD

**Issue by theseyi** [machty/ember-concurrency-decorators#31]
I'm seeing the error below when trying to annotate a generator function that takes a required parameter. If I make the parameter optional, then all is well. However, this would make my typings inaccurate!
If I define a class with the ffg attributes, I receive the commented out error from tsc.

It seems like a non-compliant type definitions, but my local investigation does not seem to back that up. Is this a known issue? How can I resolve this

```ts
export class EmberConcurrencyDecorators extends Controller {
  @task // Error: TS2345: Argument of type '"approve"' is not assignable to parameter of type '"model" | "_super" | undefined'.
  *approve(id: string) {
    yield id
  }


  @task // OK!
  *disapprove(id?: string) {
    yield id
  }
}
```

**Comment by simonihmig** [comment]
Seeing the same!

My dependencies (excerpt):
```
    "@ember-decorators/babel-transforms": "^3.1.0",
    "ember-cli-babel": "^7.2.0",
    "ember-cli-typescript": "^2.0.0-rc.2",
    "ember-decorators": "^5.0.1",
    "typescript": "^3.2.2",
```

**Comment by buschtoens** [comment]
Thanks for reporting and confirming. I added [`index.d.ts`](https://github.com/machty/ember-concurrency-decorators/blob/master/index.d.ts) in an attempt to provide as much type safety as possible. But evidently TypeScript was smarter than me.

You're doing everything correctly. I'll try to revise the types and loosen them, if not possible.

**Comment by buschtoens** [comment]
Can you share your `tsconfig.json`? I am not able to reproduce this. 🤔

**Comment by simonihmig** [comment]
Sure:

```
{
  "compilerOptions": {
    "target": "es2017",
    "allowJs": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "experimentalDecorators": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "noImplicitReturns": true,
    "noEmitOnError": false,
    "noEmit": true,
    "inlineSourceMap": true,
    "inlineSources": true,
    "baseUrl": ".",
    "module": "es6",
    "skipLibCheck": true,
    "paths": {
      "mpp-frontend/tests/*": [
        "tests/*"
      ],
      "mpp-frontend/mirage/*": [
        "mirage/*"
      ],
      "mpp-frontend/src/*": [
        "src/*"
      ],
      "*": [
        "types/*"
      ]
    }
  },
  "include": [
    "src/**/*",
    "tests/**/*",
    "types/**/*",
    "mirage/**/*"
  ]
}
```

**Comment by theseyi** [comment]
in `base.tsconfig`

```json
{
    "compilerOptions": {
      "target": "es2017",
      "allowJs": true,
      "moduleResolution": "node",
      "allowSyntheticDefaultImports": true,
      "noImplicitAny": true,
      "noImplicitThis": true,
      "alwaysStrict": true,
      "strictNullChecks": true,
      "strictPropertyInitialization": true,
      "noFallthroughCasesInSwitch": true,
      "noUnusedLocals": true,
      "noUnusedParameters": true,
      "noImplicitReturns": true,
      "noEmitOnError": false,
      "noEmit": true,
      "inlineSourceMap": true,
      "inlineSources": true,
      "module": "es6",
      "lib": ["es2017.object", "es2016", "dom", "dom.iterable", "scripthost"],
      "experimentalDecorators": true,
      "downlevelIteration": true
    },
    "exclude": ["tmp", "dist", "vendor", ".git", ".gradle", ".idea", "logs"]
  }
```


in `extended.tsconfig`
```json
{
  "extends": "../tsconfig-base",
  "compilerOptions": {
    "baseUrl": ".",
    "alwaysStrict": false,
    "strictPropertyInitialization": false,
    "lib": ["es2017.object", "es2016", "dom", "dom.iterable", "scripthost"],
    "listEmittedFiles": false,
    "paths": {
      "app/*": ["app/*", "node_modules/<addon-app>"],
      "app/tests/*": ["tests/*"],
      "app/mirage/*": ["mirage/*"],
      "addon-app": ["node_modules/<addon-app>/addon"],
      "addon-app/*": ["node_modules/<addon-app>/addon/*"]
    }
  },
  "include": [
    "app/**/*",
    "tests/**/*",
    "mirage/**/*",
    "node_modules/<addon-app>/app",
    "node_modules/<addon-app>/addon"
  ]
}
```

**Comment by buschtoens** [comment]
Thanks a lot for reporting and providing your configs!

I published a fix as `0.5.1`. 🎉

Let me know, if it works for you. 😊

**Comment by buschtoens** [comment]
FYI you're gonna run into more type errors, when actually trying to perform the task. However, after weeks of trial and error (and despair 😅), I have found a potential solution: https://github.com/machty/ember-concurrency-decorators/issues/30

It's definitely not perfect, but I think it provides the least amount of friction for the most use cases. I've posted it to #e-typescript as well, to get some export opinions on it.
